hibernate-core/src/main/java/org/hibernate/engine/spi/LoadQueryInfluencers.java

```diff
@@ -183,18 +187,24 @@ public void disableFetchProfile(String name) throws UnknownProfileException {
 	}
 
 	public EntityGraph getFetchGraph() {
-		return fetchGraph;
+		return entityGraphSemantic == EntityGraphSemantic.FETCH
+				? entityGraph
+				: null;
 	}
 
 	public void setFetchGraph(final EntityGraph fetchGraph) {
-		this.fetchGraph = fetchGraph;
+		this.entityGraphSemantic = EntityGraphSemantic.FETCH;
```

Comment:
To an outsider setFetchGraph and setLoadGraph may sound independent but we know they aren't now. The effect is also visible via the getters. "Why did A disappear if I set B". Would it not be more "honest" and clear to just accept this way all the say and have just a "set[Entity]Graph" that takes two parameters - EntityGraph and EntityGraphSemantic and a getter for each - get[Entity]Graph and get[Entity]GraphSemantic?

Alternatively setters could be renamed to something else, such as "set**As**[Fetch|Load]Graph" but this does not help with getters.

Reply:
As I mentioned earlier on another of your comments, I agree.  Ideally this would use `EffectiveEntityGraph`.  But see my other comment wrt concerns there

Reply:
So to me, this is the biggest outstanding question.  How do we track the "currently effective EntityGraph" wrt the Session?  

1. Keep `LoadQueryInfluencers` as it is more-or-less?
2. Use `EffectiveEntityGraph` in `LoadQueryInfluencers`?
3. Another option?

What do y'all think about that?

Reply:
Just to make it clear w.r.t. language used, I do not think that load/fetch graphs should be at session granularity but only specific to the query. I also think that fetch/load graph distinction is rather unfortunate and causes less than desirable practices. The situation is what it is, though, and we have to deal with it.

A central question to this is can there be both a load and fetch graph at the same time. As discussed before, I don't favour such a case, but JP seems unclear about this so my original code allowed it, leaving it to the implementation to decide and untie the knot one way or another. However, let's consider a few possibilities:

**1. We need to support the case when both fetch and load graphs are specified**
Unlikely. But, if so, we should expose the API that allows this cleanly. If present implementation does not support this, as any good API, it should fail fast. For example, calling a setter of another after one has been already set to non-null could throw an Unsupported...Exception or IllegalStateException or something. If the implementation already supports this we need to consider whether that is desirable and then come back to this consideration.

**2. Presently we do not need both at the same time but we may have to in the future**
The API needs to be made in a way that discourages even a possibility of doing so now but also in a way that could be modified later to allow it.

**3. We don't and will never need both at the same time**
The API should be clear about this at first glance such that there can be no confusion even if someone does not read documentation. This may be subjective, but to me that either means a single setter (taking two arguments, so not a conventional setter) or method(s) that do(es) not look at all like setters, forcing the developer to read the documentation. Since "fetch" and "load" really do not at all say how they differ it would be nice to use better names too. FETCH/LOAD enums could be hidden from public completely. Intentionally verbose examples:

`overrideNeededAttributes(...)`

`addNeededAttributes(...)`

Note that "add" here is slightly different and implies a merge if called multiple times. Brainstorming: resetNeededAttributes() and resetNeededAttributes(EntityGraph additional), needDefaultAttributes(), needDefaultAttributesPlusAdditional(EntityGraph), ...

Reply:
Re (1) and (2) I agree about their (lack of) usefulness. JPA certainly does not (can not) prevent setting both and I haven't seen any text explicitly stating what the two of us agree about. However, let's not waste any more time on that.

With respect to standard JPA terms being used for this I can say that I do agree that this is indeed a fact, albeit an unfortunate and very poorly documented one, as noted above. Words "fetch" and "load" really mean the same thing. Some documentation can be found here:

https://docs.oracle.com/javaee/7/tutorial/persistence-entitygraphs001.htm

Quotes, differences emphasized:

- A **fetch** graph consists of **only** the fields explicitly specified in the EntityGraph instance, **and ignores the** default entity graph settings.
- A **load** graph consists of the fields explicitly specified in the EntityGraph instance **plus any fields** in the default entity graph.

In other words, "load" **adds** to default entity graph and "fetch" **overrides** or resets it. These were brainstorming terms only, for illustration. I appended "NeededAttributes" to avoid reusing the "fetch" or "load" words in there. 

What we need to get right is, most importantly, the API (both external and internal), to promote correct development. Internal implementation may be optimized later as well, if such optimizations are discovered and needed. I think that one graph + one enum could be sufficiently efficient from memory usage perspective. Due to likely memory alignment it would not be much or at all different from graph + graph (max one non-null) or graph + boolean (e.g. "isFetchGraph").

Another alternative to mention (but, I think, NOT to pursue), is to have Fetch and Load graphs be of different classes. In some cases this may be marginally better in terms of memory use but it complicates matters otherwise - these things are indeed equal internally. To be honest, an advantage would be that the intention that the graph was created with will be clear and using graphs for what they are not intended for could be checked at compile time.